pkg/services/ngalert/notifier/channels/email_test.go

```diff
@@ -106,3 +108,193 @@ func TestEmailNotifier(t *testing.T) {
+						Annotations: model.LabelSet{"runbook_url": "http://fix.me", "__dashboardUid__": "abc", "__panelId__": "5"},
+					},
+				},
+				{
+					Alert: model.Alert{
+						Labels:      model.LabelSet{"alertname": "FiringTwo", "severity": "critical"},
+						Annotations: model.LabelSet{"runbook_url": "http://fix.me", "__dashboardUid__": "abc", "__panelId__": "5"},
+					},
+				},
+			},
+			messageTmpl: `Hi, this is a custom template.
+				{{ if gt (len .Alerts.Firing) 0 }}
+					You have {{ len .Alerts.Firing }} alerts firing.
+					{{ range .Alerts.Firing }} Firing: {{ .Labels.alertname }} at {{ .Labels.severity }} {{ end }}
+				{{ end }}`,
+			expSubject: "[FIRING:2]  ",
+			expSnippets: []string{
+				"Hi, this is a custom template.",
+				"You have 2 alerts firing.",
+				"Firing: FiringOne at warning",
+				"Firing: FiringTwo at critical",
+			},
+		},
+		{
+			name: "empty message with alerts uses default template content",
+			alerts: []*types.Alert{
+				{
+					Alert: model.Alert{
+						Labels:      model.LabelSet{"alertname": "FiringOne", "severity": "warning"},
+						Annotations: model.LabelSet{"runbook_url": "http://fix.me", "__dashboardUid__": "abc", "__panelId__": "5"},
+					},
+				},
+				{
+					Alert: model.Alert{
+						Labels:      model.LabelSet{"alertname": "FiringTwo", "severity": "critical"},
+						Annotations: model.LabelSet{"runbook_url": "http://fix.me", "__dashboardUid__": "abc", "__panelId__": "5"},
+					},
+				},
+			},
+			messageTmpl: "",
+			expSubject:  "[FIRING:2]  ",
+			expSnippets: []string{
+				"Firing: 2 alerts",
+				"<li>alertname: FiringOne</li><li>severity: warning</li>",
+				"<li>alertname: FiringTwo</li><li>severity: critical</li>",
+				"<a href=\"http://fix.me\"",
+				"<a href=\"http://localhost/base/d/abc",
+				"<a href=\"http://localhost/base/d/abc?viewPanel=5",
+			},
+		},
+		{
+			name: "message containing HTML gets HTMLencoded",
+			alerts: []*types.Alert{
+				{
+					Alert: model.Alert{
+						Labels:      model.LabelSet{"alertname": "AlwaysFiring", "severity": "warning"},
+						Annotations: model.LabelSet{"runbook_url": "http://fix.me", "__dashboardUid__": "abc", "__panelId__": "5"},
+					},
+				},
+			},
+			messageTmpl: `<marquee>Hi, this is a custom template.</marquee>
+				{{ if gt (len .Alerts.Firing) 0 }}
+					<ol>
+					{{range .Alerts.Firing }}<li>Firing: {{ .Labels.alertname }} at {{ .Labels.severity }} </li> {{ end }}
+					</ol>
+				{{ end }}`,
+			expSubject: "[FIRING:1]  (AlwaysFiring warning)",
+			expSnippets: []string{
+				"&lt;marquee&gt;Hi, this is a custom template.&lt;/marquee&gt;",
+				"&lt;li&gt;Firing: AlwaysFiring at warning &lt;/li&gt;",
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			emailNotifier := createSut(t, c.messageTmpl, emailTmpl, bus)
+
+			ok, err := emailNotifier.Notify(context.Background(), c.alerts...)
+			require.NoError(t, err)
```

Comment:
I can add these, but they'd need to be a separate test altogether. This describes a unit test of the email channel, rather than an integration test of the entire `Alerting email -> Core email` subsystem, since in these cases the notification would error out completely before ever managing to talk to the core emailer.

Also, note that the goal of this PR is not to gain 100% coverage. Instead, it's to make it possible to write integration tests spanning the `ngalert` and `email` packages, plus some baseline tests for common workflows, to be expanded upon in the future.

Reply:
If you're okay with it, I'm going to skip this for now, since this PR is more focused on test infrastructure (making it possible to write the tests we want) rather than coverage There are follow-up PRs that will increase coverage anyway. If we feel really strongly about these tests I'm still happy to add them, but I also want to balance it against getting this in soon.